presto-common/src/main/java/com/facebook/presto/common/block/ColumnarRow.java

```diff
@@ -0,0 +1,219 @@
+        Block[] fieldBlocks = new Block[rowBlock.numFields];
+        for (int i = 0; i < fieldBlocks.length; i++) {
+            fieldBlocks[i] = rowBlock.getRawFieldBlocks()[i].getRegion(firstRowPosition, totalRowCount);
+        }
+
+        return new ColumnarRow(block, fieldBlocks, block.getRetainedSizeInBytes(), block.getSizeInBytes());
+    }
+
+    private static ColumnarRow toColumnarRow(DictionaryBlock dictionaryBlock)
+    {
+        // build a mapping from the old dictionary to a new dictionary with nulls removed
+        Block dictionary = dictionaryBlock.getDictionary();
+        int[] newDictionaryIndex = new int[dictionary.getPositionCount()];
+        int nextNewDictionaryIndex = 0;
+        for (int position = 0; position < dictionary.getPositionCount(); position++) {
+            if (!dictionary.isNull(position)) {
+                newDictionaryIndex[position] = nextNewDictionaryIndex;
+                nextNewDictionaryIndex++;
+            }
+        }
+
+        // reindex the dictionary
+        int positionCount = dictionaryBlock.getPositionCount();
+        int[] dictionaryIds = new int[positionCount];
+        int nonNullPositionCount = 0;
+        for (int position = 0; position < positionCount; position++) {
+            if (!dictionaryBlock.isNull(position)) {
+                int oldDictionaryId = dictionaryBlock.getId(position);
+                dictionaryIds[nonNullPositionCount] = newDictionaryIndex[oldDictionaryId];
+                nonNullPositionCount++;
+            }
+        }
+
+        ColumnarRow columnarRow = toColumnarRow(dictionary);
+        Block[] fields = new Block[columnarRow.getFieldCount()];
+        long averageRowSize = 0;
+        for (int i = 0; i < columnarRow.getFieldCount(); i++) {
+            Block field = columnarRow.getField(i);
+            fields[i] = new DictionaryBlock(nonNullPositionCount, field, dictionaryIds);
+            averageRowSize += field.getPositionCount() == 0 ? 0 : field.getSizeInBytes() / field.getPositionCount();
+        }
+        return new ColumnarRow(
+                dictionaryBlock,
+                fields,
+                INSTANCE_SIZE + dictionaryBlock.getRetainedSizeInBytes() + sizeOf(dictionaryIds),
+                // The estimated serialized size is the sum of the following:
+                // 1) the offsets size: Integer.BYTES * positionCount. Note that even though ColumnarRow doesn't have the offsets array, the serialized RowBlock still has it. Please see RowBlockEncodingBuffer.
+                // 2) nulls array size: Byte.BYTES * positionCount
+                // 3) the estimated serialized size for the fields Blocks which were just constructed as new DictionaryBlocks:
+                //     the average row size: averageRowSize * the number of rows: nonNullPositionCount
+                (Integer.BYTES + Byte.BYTES) * positionCount + averageRowSize * nonNullPositionCount);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).